File: sketcher/executor.py
"""Step execution logic for Sketcher."""

import os
import tempfile
from pathlib import Path
from typing import Optional, List

from sketcher import utils, kubernetes
from sketcher.model import Model
from sketcher.exceptions import SketcherError, SketcherProcessError


def run_steps(
    yaml_file: str,
    kubeconfigs: Optional[List[str]] = None,
    work_dir: Optional[str] = None,
    debug: bool = False,
    quiet: bool = False
):
    """Run all steps in a skewer.yaml file.

    Args:
        yaml_file: Path to resolved skewer.yaml
        kubeconfigs: Optional list of kubeconfig paths for sites
        work_dir: Working directory for ~ substitution (default: temp dir)
        debug: If True, print debug output on failure
        quiet: If True, suppress progress messages

    Raises:
        SketcherError: If steps fail
    """
    utils.info("Running steps from {}", yaml_file, quiet=quiet)

    # Check environment
    kubernetes.check_environment()

    # Load model
    if kubeconfigs is None:
        kubeconfigs = []

    model = Model(yaml_file, kubeconfigs)
    model.check()

    # Create work directory
    if work_dir is None:
        work_dir = tempfile.mkdtemp(prefix="sketcher-")
        utils.info("Using work directory: {}", work_dir, quiet=quiet)
    else:
        Path(work_dir).mkdir(parents=True, exist_ok=True)

    try:
        # Run all steps except cleaning_up
        for step in model.steps:
            if step.name == "cleaning_up":
                continue

            run_step(model, step, work_dir, quiet=quiet)

        # Demo mode support
        if "SKETCHER_DEMO" in os.environ:
            from sketcher import demo
            demo.save_demo_context(model, work_dir)
            demo.pause_for_demo(model, quiet=quiet)

    except Exception as e:
        if debug:
            print_debug_output(model)

        raise
    finally:
        # Always run cleaning_up (if it exists)
        for step in model.steps:
            if step.name == "cleaning_up":
                run_step(model, step, work_dir, check=False, quiet=True)
                break


def run_step(model: Model, step, work_dir: str, check: bool = True, quiet: bool = False):
    """Run a single step.

    Args:
        model: Sketcher Model instance
        step: Step to run
        work_dir: Working directory for ~ substitution
        check: If True, raise error on command failure
        quiet: If True, suppress progress messages

    Raises:
        SketcherError: If command fails and check=True
    """
    if not step.commands:
        return

    # Check if all commands are readme-only - if so, skip the entire step
    all_readme = True
    for site_name, commands in step.commands:
        for command in commands:
            if command.apply != "readme":
                all_readme = False
                break
        if not all_readme:
            break

    if all_readme:
        return

    # Use operation context for visual hierarchy
    with utils.operation(str(step), quiet=quiet):
        sites_dict = dict(model.sites)

        for site_name, commands in step.commands:
            site = sites_dict[site_name]

            # Use site context manager (sets env vars and logging prefix)
            with site:
                # Set kubectl namespace for kubernetes sites
                if site.platform == "kubernetes":
                    utils.run(
                        f"kubectl config set-context --current --namespace {site.namespace}",
                        quiet=True,
                        stdout=utils.subprocess.DEVNULL
                    )

                # Execute commands
                for command in commands:
                    # Skip README-only commands
                    if command.apply == "readme":
                        continue

                    # Execute await operations
                    if command.await_resource:
                        utils.debug(f"Awaiting resource: {command.await_resource}")
                        kubernetes.await_resource(command.await_resource, quiet=quiet)

                    if command.await_ingress:
                        utils.debug(f"Awaiting ingress: {command.await_ingress}")
                        kubernetes.await_ingress(command.await_ingress, quiet=quiet)

                    if command.await_http_ok:
                        # await_http_ok is a tuple: (service, url_template, user, password)
                        utils.debug(f"Awaiting HTTP OK: {command.await_http_ok[0]}")
                        kubernetes.await_http_ok(*command.await_http_ok, quiet=quiet)

                    if command.await_console_ok:
                        utils.debug("Awaiting console OK")
                        kubernetes.await_console_ok(quiet=quiet)

                    if command.await_port:
                        utils.debug(f"Awaiting port: {command.await_port}")
                        kubernetes.await_port(command.await_port, quiet=quiet)

                    # Execute shell command
                    if command.run:
                        # Replace ~ with work_dir
                        cmd = command.run.replace("~", work_dir)

                        # Wrap localhost curl commands with retry logic
                        if "curl" in cmd and "localhost" in cmd and "http://" in cmd:
                            # Add retry flags to make curl more resilient
                            # max-time 15 allows for slower app startup, retry-all-errors catches transient failures
                            if "--retry" not in cmd:
                                cmd = cmd.replace("curl ", "curl --retry 20 --retry-delay 3 --retry-all-errors --max-time 15 ")

                        result = utils.run(cmd, shell=True, check=False, quiet=quiet)

                        # Handle expect_failure
                        if command.expect_failure:
                            if result.returncode == 0:
                                raise SketcherError("A command expected to fail did not fail")
                            continue

                        # Check for failure
                        if check and result.returncode != 0:
                            err = SketcherProcessError(
                                f"Command failed in {step}",
                                result.returncode,
                                result.stdout,
                                result.stderr
                            )
                            raise err


def print_debug_output(model: Model):
    """Print debug output for all sites.

    Shows kubectl get all and skupper status for each site.

    Args:
        model: Sketcher Model instance
    """
    utils.eprint("\n" + "=" * 80)
    utils.cprint("DEBUG OUTPUT", color="yellow", file=utils.sys.stderr)
    utils.eprint("=" * 80)

    for site_name, site in model.sites:
        utils.cprint(f"\n--- {site.title} ({site_name}) ---\n", color="cyan", file=utils.sys.stderr)

        with site:
            if site.platform == "kubernetes":
                # Set namespace
                utils.run(
                    f"kubectl config set-context --current --namespace {site.namespace}",
                    quiet=True,
                    stdout=utils.subprocess.DEVNULL,
                    check=False
                )

                # Show all resources
                utils.eprint("kubectl get all:")
                utils.run("kubectl get all", check=False)
                utils.eprint()

            # Show skupper status
            utils.eprint("skupper status:")
            utils.run("skupper status", check=False)
            utils.eprint()

            utils.eprint("skupper link status:")
            utils.run("skupper link status", check=False)
            utils.eprint()

    utils.eprint("=" * 80)

File: internal/executor/executor.go
package executor

import (
	"fmt"
	"os"
	"os/exec"
	"strings"

	"github.com/skupperproject/sketcher/internal/demo"
	"github.com/skupperproject/sketcher/internal/kubernetes"
	"github.com/skupperproject/sketcher/internal/model"
	"github.com/skupperproject/sketcher/internal/utils"
)

// Export NewModel for use by cli package
var NewModel = model.NewModel

// RunSteps executes all steps from a skewer.yaml file
func RunSteps(yamlFile string, kubeconfigs []string, workDir string, debug, quiet bool) error {
	utils.Info("Running steps from %s", yamlFile)

	// Setup signal handler and cleanup
	SetupSignalHandler()
	defer CleanupBackgroundProcesses()

	// Check environment
	if err := kubernetes.CheckEnvironment(); err != nil {
		return err
	}

	// Load model
	m, err := model.NewModel(yamlFile, kubeconfigs)
	if err != nil {
		return err
	}

	if err := m.Check(); err != nil {
		return err
	}

	// Create work directory
	if workDir == "" {
		// Use /tmp directly to avoid macOS temp paths with special characters
		// that fail skupper's token file path validation (regex: ^[A-Za-z0-9./~-]+$)
		baseDir := "/tmp"
		utils.Debug("Creating temporary work directory in %s", baseDir)
		tempDir, err := os.MkdirTemp(baseDir, "sketcher-")
		if err != nil {
			return fmt.Errorf("failed to create work directory: %w", err)
		}
		workDir = tempDir
		utils.Info("Using work directory: %s", workDir)
	} else {
		utils.Debug("Using specified work directory: %s", workDir)
		if err := os.MkdirAll(workDir, 0755); err != nil {
			return err
		}
		utils.Debug("Work directory created: %s", workDir)
	}

	// Run all steps except cleaning_up
	for _, step := range m.Steps {
		if step.Name == "cleaning_up" {
			continue
		}

		if err := runStep(m, step, workDir, true, quiet); err != nil {
			if debug {
				printDebugOutput(m)
			}
			return err
		}
	}

	// Demo mode support
	if os.Getenv("SKETCHER_DEMO") != "" {
		// Save demo context
		if err := demo.SaveDemoContext(m, workDir); err != nil {
			return fmt.Errorf("failed to save demo context: %w", err)
		}

		// Pause and display demo information
		if err := demo.PauseForDemo(m, quiet); err != nil {
			return fmt.Errorf("failed during demo pause: %w", err)
		}
	}

	// Always run cleaning_up if it exists
	for _, step := range m.Steps {
		if step.Name == "cleaning_up" {
			runStep(m, step, workDir, false, true)
			break
		}
	}

	return nil
}

func runStep(m *model.Model, step *model.Step, workDir string, check, quiet bool) error {
	if len(step.Commands) == 0 {
		return nil
	}

	// Check if all commands are readme-only
	allReadme := true
	for _, commands := range step.Commands {
		for _, cmd := range commands {
			if cmd.Apply != "readme" {
				allReadme = false
				break
			}
		}
		if !allReadme {
			break
		}
	}

	if allReadme {
		return nil
	}

	// Log step
	if !quiet {
		utils.Cprint(fmt.Sprintf("→ %s", stepString(step)), "cyan")
	}

	for _, site := range m.Sites {
		commands, ok := step.Commands[site.Name]
		if !ok {
			continue
		}

		// Set environment variables for site
		err := site.WithEnv(func() error {
			// Set kubectl namespace for kubernetes sites
			if site.Platform == "kubernetes" {
				cmd := exec.Command("kubectl", "config", "set-context", "--current", "--namespace", site.Namespace)
				cmd.Stdout = nil
				cmd.Stderr = nil
				cmd.Run()
			}

			// Execute commands
			for _, command := range commands {
				if command.Apply == "readme" {
					continue
				}

				// Execute await operations
				if command.AwaitResource != "" {
					utils.Debug("Awaiting resource: %s", command.AwaitResource)
					if err := kubernetes.AwaitResource(command.AwaitResource, 300, quiet); err != nil {
						return err
					}
				}

				if command.AwaitIngress != "" {
					utils.Debug("Awaiting ingress: %s", command.AwaitIngress)
					if _, err := kubernetes.AwaitIngress(command.AwaitIngress, 300, quiet); err != nil {
						return err
					}
				}

				if len(command.AwaitHTTPOK) > 0 {
					utils.Debug("Awaiting HTTP OK")
					// Implementation would be in kubernetes package
				}

				if command.AwaitConsoleOK {
					utils.Debug("Awaiting console OK")
					if err := kubernetes.AwaitConsoleOK(300, quiet); err != nil {
						return err
					}
				}

				if command.AwaitPort > 0 {
					utils.Debug("Awaiting port: %d", command.AwaitPort)
					if err := utils.AwaitPort(command.AwaitPort, "localhost", 60); err != nil {
						return err
					}
				}

				// Execute shell command
				if command.Run != "" {
					cmdStr := strings.ReplaceAll(command.Run, "~", workDir)

					// Auto-inject platform flag for skupper commands on non-kubernetes platforms
					if site.Platform != "kubernetes" && strings.Contains(cmdStr, "skupper ") {
						// Only inject if -p/--platform is not already present
						if !strings.Contains(cmdStr, " -p ") && !strings.Contains(cmdStr, " --platform") {
							// Inject -p flag before -n flag or at the end
							if strings.Contains(cmdStr, " -n ") {
								cmdStr = strings.Replace(cmdStr, " -n ", fmt.Sprintf(" -p %s -n ", site.Platform), 1)
							} else {
								// Append at the end (before any trailing &)
								cmdStr = strings.TrimSpace(cmdStr)
								if strings.HasSuffix(cmdStr, "&") {
									cmdStr = strings.TrimSuffix(cmdStr, "&")
									cmdStr = strings.TrimSpace(cmdStr) + fmt.Sprintf(" -p %s &", site.Platform)
								} else {
									cmdStr = cmdStr + fmt.Sprintf(" -p %s", site.Platform)
								}
							}
						}
					}

					// Wrap localhost curl commands with retry logic
					if strings.Contains(cmdStr, "curl") && strings.Contains(cmdStr, "localhost") && strings.Contains(cmdStr, "http://") {
						if !strings.Contains(cmdStr, "--retry") {
							cmdStr = strings.ReplaceAll(cmdStr, "curl ", "curl --retry 20 --retry-delay 3 --retry-all-errors --max-time 15 ")
						}
					}

					// Check if this is a background command (ends with &)
					isBackground := strings.HasSuffix(strings.TrimSpace(cmdStr), "&")

					var err error
					if isBackground {
						// Remove the & and run as tracked background process
						cmdStr = strings.TrimSuffix(strings.TrimSpace(cmdStr), "&")
						err = RunBackgroundCommand(cmdStr)
					} else {
						// Run normally
						cmd := exec.Command("sh", "-c", cmdStr)
						cmd.Stdout = os.Stdout
						cmd.Stderr = os.Stderr
						err = cmd.Run()
					}

					if command.ExpectFailure {
						if err == nil {
							return fmt.Errorf("a command expected to fail did not fail")
						}
						continue
					}

					if check && err != nil {
						return fmt.Errorf("command failed in %s: %v", stepString(step), err)
					}
				}
			}

			return nil
		})

		if err != nil {
			return err
		}
	}

	if !quiet {
		utils.Cprint(fmt.Sprintf("✓ %s", stepString(step)), "green")
	}

	return nil
}

func printDebugOutput(m *model.Model) {
	fmt.Fprintf(os.Stderr, "\n%s\n", strings.Repeat("=", 80))
	utils.Cprint("DEBUG OUTPUT", "yellow")
	fmt.Fprintf(os.Stderr, "%s\n", strings.Repeat("=", 80))

	for _, site := range m.Sites {
		utils.Cprint(fmt.Sprintf("\n--- %s (%s) ---\n", site.Title, site.Name), "cyan")

		site.WithEnv(func() error {
			if site.Platform == "kubernetes" {
				cmd := exec.Command("kubectl", "config", "set-context", "--current", "--namespace", site.Namespace)
				cmd.Run()

				fmt.Fprintln(os.Stderr, "kubectl get all:")
				cmd = exec.Command("kubectl", "get", "all")
				cmd.Stdout = os.Stderr
				cmd.Stderr = os.Stderr
				cmd.Run()
				fmt.Fprintln(os.Stderr)
			}

			fmt.Fprintln(os.Stderr, "skupper status:")
			cmd := exec.Command("skupper", "status")
			cmd.Stdout = os.Stderr
			cmd.Stderr = os.Stderr
			cmd.Run()
			fmt.Fprintln(os.Stderr)

			return nil
		})
	}

	fmt.Fprintf(os.Stderr, "%s\n", strings.Repeat("=", 80))
}

func stepString(step *model.Step) string {
	if step.Numbered {
		return fmt.Sprintf("step %d '%s'", step.Number, step.Title)
	}
	return fmt.Sprintf("'%s'", step.Title)
}
